vignettes/datatable-benchmarking.Rmd

```diff
@@ -79,42 +105,83 @@ options(datatable.optimize=3L)
 
-# avoid class coercion
+### avoid class coercion
 
-Unless this is what you truly want to measure you should prepare input objects of the expected class for every tool you are benchmarking.
+Unless this is what you truly want to measure, you should prepare input objects of the expected class for every tool you are benchmarking so that the benchmark can measure the timing of the actual relevant computation rather than class coercion time.
 
-# avoid `microbenchmark(..., times=100)`
+### avoid `microbenchmark(..., times=100)`
 
+Be sure to read _General suggestions_ section in the top of this document which also covers this topic. 
 Repeating a benchmark many times usually does not give the clearest picture for data processing tools. Of course, it makes perfect sense for more atomic calculations, but this is not a good representation of the most common way these tools will actually be used, namely for data processing tasks, which consist of batches of sequentially provided transformations, each run once.
+
 Matt once said:
 
 > I'm very wary of benchmarks measured in anything under 1 second. Much prefer 10 seconds or more for a single run, achieved by increasing data size. A repetition count of 500 is setting off alarm bells. 3-5 runs should be enough to convince on larger data. Call overhead and time to GC affect inferences at this very small scale.
 
-This is very valid. The smaller time measurement is the relatively bigger noise is. Noise generated by method dispatch, package/class initialization, etc. Main focus of benchmark should be on real use case scenarios.
+This is very valid. The smaller time measurement is the relatively bigger noise is. Noise generated by method dispatch, package/class initialization, etc. Main focus of benchmark should be real use case scenarios.
+
+Example below represents the problem discussed:
+```r
+library(microbenchmark)
+library(data.table)
+set.seed(108)
+
+N = 1e5L
+dt = data.table(id=sample(N), value=rnorm(N))
+setindex(dt, "id")
+df = as.data.frame(dt)
+microbenchmark(
+  dt[id==5e4L, value],
+  df[df$id==5e4L, "value"],
+  times = 1000
+)
+#Unit: microseconds
+#                        expr      min        lq      mean    median        uq       max neval
+#     dt[id == 50000L, value] 1237.964 1359.5635 1466.9513 1392.1735 1443.1725 14500.751  1000
+#df[df$id == 50000L, "value"]  355.063  391.2695  430.3884  404.7575  429.5605  2481.442  1000
+
+N = 1e7L
+dt = data.table(id=sample(N), value=rnorm(N))
+setindex(dt, "id")
+df = as.data.frame(dt)
+microbenchmark(
+  dt[id==5e6L, value],
```

Comment:
Why `5e6` here, I was distracted by the difference, shouldn't `5e4L` continue to demonstrate the point?